consensus/obcpbft/pbft-core.go

```diff
@@ -896,18 +922,21 @@ func (instance *pbftCore) Checkpoint(seqNo uint64, id []byte) {
 
 // execDone is an event telling us that the last execution has completed
 func (instance *pbftCore) execDone() {
-	instance.execCompleteChan <- struct{}{}
+	instance.manager.queue() <- execDoneEvent{}
 }
 
 func (instance *pbftCore) execDoneSync() {
-	logger.Info("Replica %d finished execution %d, trying next", instance.id, *instance.currentExec)
-
-	instance.lastExec = *instance.currentExec
-	instance.currentExec = nil
+	if instance.currentExec != nil {
+		logger.Info("Replica %d finished execution %d, trying next", instance.id, *instance.currentExec)
+		instance.lastExec = *instance.currentExec
+		if instance.lastExec%instance.K == 0 {
+			instance.Checkpoint(instance.lastExec, instance.consumer.getState())
+		}
 
-	if instance.lastExec%instance.K == 0 {
-		instance.Checkpoint(instance.lastExec, instance.consumer.getState())
+	} else {
```

Comment:
This will likely cause a merge conflict.  I had originally cherry picked this fix out of this branch to get it merged earlier, and then @corecode had had some feedback on it.  Ideally, I'd prefer to merge this PR first and fix up the conflict in #1596 after.